tasks/nunjucks.js

```diff
@@ -6,65 +6,159 @@
+                    return reject('For some destinations were not specified source files.')
                 }
-                grunt.file.write(f.dest, res);
-                grunt.log.writeln('File "' + f.dest + '" created.');
-                done();
-            });
-
-        }, function(err) {
-            if (err) {
-                grunt.log.error(err);
-                grunt.fail.warn('Something went wrong.');
+
+                // Iterate over files' sources
+                file.src.forEach(src => {
+                    // Сheck whether source file exists
+                    if (!grunt.file.exists(src)) {
+                        grunt.log.error(`Source file ${chalk.cyan(src)} for ${chalk.cyan(filedest)} not found.`)
+
+                        // Skip to next source file — nothing we can do with non-existing file
+                        return reject('Some source files were not found.')
+                    }
+
+                    // Construct absolute path to file for Nunjucks
+                    let filepath = path.join(process.cwd(), src)
+
+                    let data = {}
+                    // Work with data only if there is any data
+                    if (options.data) {
+                        // Clone data
+                        for (let i in options.data) {
+                            if (options.data.hasOwnProperty(i)) {
+                                data[i] = options.data[i]
+                            }
+                        }
+
+                        // Preprocess data
+                        if (typeof options.preprocessData === 'function') {
+                            data = options.preprocessData.call(file, data)
+                        }
+                    }
+
+                    // Asynchronously render templates with configurated Nunjucks environment
+                    // and write to destination
+                    env.render(filepath, data, (error, result) => {
+                        // Catch errors, warn
+                        if (error) {
+                            grunt.log.error(error)
+                            grunt.fail.warn('Failed to compile one of the source files.')
+                            grunt.log.writeln()
+
+                            // Prevent writing of failed to compile file, skip to next file
+                            return reject('Failed to compile some source files.')
+                        }
+
+                        // Write rendered template to destination
+                        grunt.file.write(filedest, result)
+
+                        // Debug process
+                        grunt.verbose.ok(`File ${chalk.cyan(filedest)} created.`)
+                        grunt.verbose.writeln()
+
+                        countCompiled++
+                    })
+                })
+            })
+
+            // Finish Promise
+            resolve()
+        })
+
+        // Print any errors from rejects
+        .catch(error => {
```

Comment:
It seems to be the only tricky part — still can't make it to work like we need it.

When promise receives `reject()`, it jumps to `.catch()` part.

If we will simulate error (no src specified), we will get this:

![1](https://cloud.githubusercontent.com/assets/4460311/10095945/232e593a-6376-11e5-9a22-2315c775aa40.png)

In other words, it jumps to `catch()`, and `.then()` immediately executed _after_ `catch()` (which is quite logically). In result we have final results in wrong place and with wrong counter value (because it was printed before Promise finished).

if we'll move `.catch()` after `.then()`, we will never receive final results at all:

![2](https://cloud.githubusercontent.com/assets/4460311/10095991/9ff118ae-6376-11e5-9054-a6d7d66cff20.png)

Maybe wrapping `Promise()` into another `PromIse().then()` will help, but it doesn't looks like a good solution.

Any ideas?